libraries/lib-ipc/IPCClient.cpp

```diff
@@ -0,0 +1,77 @@
+#include "IPCClient.h"
+#include "IPCChannel.h"
+
+#include <thread>
+#include <atomic>
+#include <stdexcept>
+
+#include "internal/ipc-types.h"
+#include "internal/socket_guard.h"
+#include "internal/BufferedIPCChannel.h"
+
+class IPCClient::Impl
+{
+   std::atomic<bool> mAlive{true};
+   std::unique_ptr<std::thread> mWorker;
+public:
+
+   ~Impl()
+   {
+      mAlive = false;
+      if(mWorker)
+         mWorker->join();
+   }
+
+   void CreateChannel(int id, IPCChannelStatusCallback* callback)
+   {
+      if(mWorker)
+         throw std::runtime_error("cannot handle more than one channel");
+
+      mWorker = std::make_unique<std::thread>([this, id, callback]()
+      {
+         auto fd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+         if(fd == INVALID_SOCKET)
+         {
+            callback->OnConnectionError();
+            return;
+         }
+
+         socket_guard fd_guard { fd };
+
+         sockaddr_in addrin {};
+         addrin.sin_family = AF_INET;
+         addrin.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+         addrin.sin_port = htons(static_cast<u_short>(id));
+
+         if(connect(fd, reinterpret_cast<const sockaddr*>(&addrin), sizeof(addrin)) != 0)
+         {
+            callback->OnConnectionError();
+            return;
+         }
+
+         BufferedIPCChannel channel(fd, callback);
+         while(mAlive && channel.Serve())
+            std::this_thread::yield();
+      });
+   }
+};
+
+IPCClient::IPCClient()
+{
+#ifdef _WIN32
+   WSADATA wsaData;
+   auto result = WSAStartup(MAKEWORD(2, 2), &wsaData);
+   if (result == NO_ERROR)
+      mImpl = std::make_unique<Impl>();
```

Comment:
On Windows this is a possibly-failing constructor but there is no indication to the caller of the failure.  Either you need a boolean method to test for success, or should throw an exception.

Same applies to the server, on Windows only.

Reply:
... Oh I see,  but then CreateChannel() can indicate error.  OK then.